src/car_reader.rs

```diff
@@ -331,6 +128,301 @@ fn ngx_buf_remove_start(buf: *mut ngx_buf_s, len: usize) {
+            // end of the frame
+            } else if current.len() >= self.len {
+                match self.state {
+                    FrameType::CarHeader => {
+                        self.state = FrameType::Block;
+                        pos += self.len;
+                    }
+                    FrameType::PBLinks => {
+                        self.state = FrameType::MerkleDag;
+                        self.blk_pos += self.len;
+                    }
+                    FrameType::UnixFsData => {
+                        self.blk_pos = 0;
+                        self.state = FrameType::Block;
+                        pos += self.blk_len;
+                    }
+                    _ => {}
+                };
+                current = &current[self.len..];
+                self.len = 0;
+
+            // partial frame
+            } else {
+                match self.state {
+                    FrameType::PBLinks => {
+                        self.blk_pos += current.len();
+                    }
+                    _ => {}
+                };
+                self.len -= current.len();
+                current = &[];
+            }
+        }
+        Ok((0, pos))
+    }
+
+    fn decode_len(&mut self, buf: &[u8]) -> Option<(usize, usize)> {
+        println!("decode_len");
+        let mut i = 0;
+        loop {
+            self.buf.push(buf[i]);
+            match usize::decode_var(&self.buf[..]) {
+                Some((size, _)) => {
+                    self.buf.clear();
+                    return Some((size, i + 1));
+                }
+                None => {
+                    if buf.len() - (i + 1) > 0 {
+                        i += 1;
+                        continue;
+                    } else {
+                        return None;
+                    }
+                }
+            };
+        }
+    }
+
+    fn decode_cid(&mut self, buf: &[u8]) -> Option<(Cid, usize)> {
+        println!("decode_cid");
+        let mut i = 0;
+
+        let filled = self.buf.len();
+
+        loop {
+            for j in i..std::cmp::min(i + 36, buf.len()) {
+                self.buf.push(buf[j]);
+                i = j;
+            }
+            // start from the next index
+            i += 1;
+            let mut reader = Cursor::new(&self.buf[..]);
+            match Cid::read_bytes(&mut reader) {
+                Ok(cid) => {
+                    let read = reader.position() as usize;
+                    self.buf.clear();
+                    self.blk_pos += read;
+                    return Some((cid, read - filled));
+                }
+                Err(_) => {
+                    if buf.len() > (i + 1) {
+                        continue;
+                    } else {
+                        return None;
+                    }
+                }
+            };
+        }
+    }
+
+    fn is_seek(&self) -> bool {
+        lt_bound(self.range.start_bound(), self.unixfs_read as u64)
+    }
```

Comment:
It looks like `is_seek` isn't used, hence can be removed. Then also `lt_bound` can be removed.